### ISSUE TITLE
size mismatch

### ISSUE THREAD

**Issue by Gbau08** [LaceyChen17/DREAM#2]
Hello,

Thank you for making your code public,I've been trying to run dream, however I keep having the error 

"size mismatch, m1: [1 x 256], m2: [64 x 256] at c:\anaconda2\conda-bld\pytorch_1513133520683\work\torch\lib\th\generic/THTensorMath.c:1416"

After examination, it comes from that part of the code:
for i,x in enumerate(batchify(train_ub, dr_config.batch_size)):
        baskets, lens, _ = x
        dr_hidden = repackage_hidden(dr_hidden) # repackage hidden state for RNN
        dr_model.zero_grad() # optim.zero_grad()
        **dynamic_user, _  = dr_model(baskets, lens, dr_hidden)**

- dr_hidden is of size 1x4x64
- baskets is a list with index from 0 to 255, each having a size of 99 (padded)
- lens is a list with index from 0 to 255, each having size dependent on the basket of the user (max at 99)

The parameter I used for the config are:

DREAM_CONFIG = {'basket_pool_type': 'max', # 'avg'
                'rnn_layers': 2, # 2, 3
                'rnn_type': 'LSTM',#'RNN_TANH',#'GRU',#'LSTM',# 'RNN_RELU',
                'dropout': 0.5,
                # 'num_product': 49688 + 1, # padding idx = 0
                'num_product': 49688 + 1 + 1, 
                # 49688 products, padding idx = 0, none idx = 49689, none idx indicates no products
                'none_idx': 49689,
                'embedding_dim': 64, # 128 
                'cuda': False, # True,
                'clip': 20, # 0.25
                'epochs': 100,
                'batch_size': 256,
                'learning_rate': 0.01, # 0.0001
                'log_interval': 1, # num of batchs between two logging
                'checkpoint_dir': DREAM_MODEL_DIR + 'reorder-next-dream-{epoch:02d}-{loss:.4f}.model',
                }

I dont use ub_rbks, ub_ihis; only ub_basket

Thanks for the help !

**Comment by yihong-chen** [comment]
Hi, Could you find the exact line in DreamModel which generates the bug i.e. which operation causes the size mismatch problem? And I'll check it further. I mainly tested dream on the reorder prediction problem but I remember that it ran well on the order prediction problem.

**Comment by Gbau08** [comment]
Yes, it comes from the function forward in class dream. More precisely, it comes from that line:
output, h_u = self.rnn(packed_ub_seqs, hidden).  

I'll try to run twith the reorder prediction to see if this persists or not.

**Comment by yihong-chen** [comment]
Could you print out the sizes of `hidden` and `ub_seqs` just before line 
    output, h_u = self.rnn(packed_ub_seqs, hidden) ?

**Comment by Gbau08** [comment]
I have the same problem with the reorder.

hidden (2 Variables)
[torch.FloatTensor of size 3x32x128]
![image](https://user-images.githubusercontent.com/20928835/37146157-415dec04-22c3-11e8-891f-650a28b1d0e4.png)

ub_seqs
torch.FloatTensor of size 32x7936
![image](https://user-images.githubusercontent.com/20928835/37146187-54318aca-22c3-11e8-8e1c-e11579f3f826.png)

packed_ub_seqs(1 Variable + 1 list of size 62)
[torch.FloatTensor of size 453]
, batch_sizes=[32, 32, 32, 26, 25, 22, 21, 18, 16, 12, 12, 12, 11, 9, 8, 8, 8, 8, 8, 7, 6, 6, 6, 5, 5, 5, 5, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1])
![image](https://user-images.githubusercontent.com/20928835/37146484-2c3195be-22c4-11e8-84fb-8c3bf4841e1c.png)

32 is the batch size, 128 the embedded dim

The error :
![image](https://user-images.githubusercontent.com/20928835/37146662-b021ebd0-22c4-11e8-8b2a-d7415241390a.png)

**Comment by yihong-chen** [comment]
Is something wrong with the ub_seqs ? Its size should be (batch_size, max_len, embedding_dim)

**Comment by Gbau08** [comment]
ub_seqs = [] # users' basket sequence
        for user in x: # x shape (batch of user, time_step, indice of product) nested lists
            embed_baskets = []
            for basket in user:
                basket = torch.LongTensor(basket).resize_(1, len(basket))
                basket = basket.cuda() if self.config.cuda else basket # use cuda for acceleration
                basket = self.encode(torch.autograd.Variable(basket)) # shape: 1, len(basket), embedding_dim
                embed_baskets.append(self.pool(basket, dim = 1))
            # concat current user's all baskets and append it to users' basket sequence
            ub_seqs.append(torch.cat(embed_baskets, 1)) # shape: 1, num_basket, embedding_dim


basket is of size 1x1x128 but embed_basket is of size 1x128, could it come from that line and the pool function?
embed_baskets.append(self.pool(basket, dim = 1))

**Comment by Gbau08** [comment]
I checked with a bastet of length 18, indeed basket has size 1x18x128 (1x length (basket) x embedded dim, but embed_baskets is a list of 74 Variables, and size is reported to be 1x128.

When basket is converted into a float tensor with self.encode, it gives me the right dimensions : 1x18x128. After the application of the pool function (the max), I got 1x128 ( like in the paper, it takes the maximum value of every dimension). 
Embedded basket is a list of  with 74 Variables/tensors (74 is the number of baskets in the batch I assume),each basket with size 1x128

torch.cat(embed_baskets, 1) however transform this into 1x9472 for each user.

**Comment by yihong-chen** [comment]
Maybe we should check size of `torch.cat(embed_baskets, 1)` ? in line 
ub_seqs.append(torch.cat(embed_baskets, 1))

**Comment by Gbau08** [comment]
Yes, it is indeed that line.
embed_baskets is a list of each baskets (74) in the batch, each basket has 1x128 dim.
If we do torch.cat(embed_baskets, 1), we end up with a long vector of 1x9472 (74x128=9472).

If I follow you, I should have instead (34,74,128) (batch_size, max_len, embedding_dim).

**Comment by Gbau08** [comment]
Problem was resolved by using stack instead of cat

            userSequence= torch.cat(embed_baskets, 0) 
            ub_seqs.append(userSequence)
            ub_seqsCat= torch.stack(ub_seqs, 0)
